src/source_sender/mod.rs

```diff
@@ -200,19 +215,45 @@ impl SourceSender {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 struct Inner {
     inner: LimitedSender<EventArray>,
     output: String,
+    lag_time: Option<Histogram>,
+}
+
+impl fmt::Debug for Inner {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Inner")
+            .field("inner", &self.inner)
+            .field("output", &self.output)
+            // `metrics::Histogram` is missing `impl Debug`
+            .finish()
+    }
 }
 
 impl Inner {
-    fn new_with_buffer(n: usize, output: String) -> (Self, LimitedReceiver<EventArray>) {
+    fn new_with_buffer(
+        n: usize,
+        output: String,
+        lag_time: Option<Histogram>,
+    ) -> (Self, LimitedReceiver<EventArray>) {
         let (tx, rx) = channel::limited(n);
-        (Self { inner: tx, output }, rx)
+        (
+            Self {
+                inner: tx,
+                output,
+                lag_time,
+            },
+            rx,
+        )
     }
 
     async fn send(&mut self, events: EventArray) -> Result<(), ClosedError> {
+        let reference = Utc::now().timestamp_millis();
+        events
+            .iter_events()
+            .for_each(|event| self.emit_lag_time(event, reference));
```

Comment:
One thing we could consider is adding a timestamp field to the `EventArray` itself. There would be consistency concerns to think about, especially if event timestamps are changed within a codec or similar, but also a potentially meaningful performance improvement.

Reply:
I'm not sure how that would work if the events composing the array have differing timestamps.